### ISSUE TITLE
High velocity player-entity collisions often fail

### ISSUE THREAD

**Issue by TheTermos** [minetest/minetest#9806]
##### Minetest version
```
All of them
```

##### OS / Hardware
```
Any
```

##### Summary
High velocity head on player - entity collisions often fail, they pass through each other. Both have to be moving in opposite directions for this to happen.

Here's what I believe is happening:
- collision works correctly on both ends, but because of update delay, results are slightly different.
- server side the entity collides farther forward, because the info about player moving is late.
- then the client receives an update about entity's position, now it's clipping with player's collision box client side.
- forward key is still being held, and now nothing is stopping player from moving farther forward. 

##### Steps to reproduce
Repeatedly collide player into an entity moving towards them head on.
It's the easiest to reproduce with fast move on, however it's still theoretically possible with normal player speed.

Note: it doesn't occur every single time only because collision detection implements clipping recovery (old as well as new).

**Comment by raymoo** [comment]
My initial thought is that this is just a consequence of player physics being client-side and Minetest trading off physics consistency for lower network bandwidth and simpler networking code. I don't think that this can be fixed without player physics being determined by the server and stronger consistency being enforced between the client and server states.

**Comment by TheTermos** [comment]
> I don't think that this can be fixed without player physics being determined by the server and stronger consistency being enforced between the client and server states.

I think fixing that foundational error is unlikely to happen, based on how ineffective the development process is, and that hardly anyone seems to care.

How about some kind of relatively simple workaround then? Just off the top of my head:
- upon receiving an update, the client attempts a move instead of setting a position.
- if the move resulted in collision with a node, set the position anyway.

That would keep entities slightly out of sync with the server as as long as there's any other cao in the way, but should prevent players from walking through them.

**Comment by orbea** [comment]
I think if temporary hacks are added they will become permanent hacks and its better to do it right to begin with.

> based on how ineffective the development process is, and that hardly anyone seems to care.

Caring is hard when minetest maintains a PR graveyard.

**Comment by TheTermos** [comment]
> I think if temporary hacks are added they will become permanent hacks and its better to do it right to begin with.

Well, a permanent hack is the idea.
In my estimates the odds of ever doing it right are virtually zero. I wouldn't mind being wrong about this though.

**Comment by raymoo** [comment]
>upon receiving an update, the client attempts a move instead of setting a position.

The client already "attempts moves" (assuming you mean checks for collision when setting position, please clarify if you mean something else) when processing player movement. 

>In my estimates the odds of ever doing it right are virtually zero.

I agree, and this isn't solely a criticism of Minetest development or anything. Server-authoritative player movement would require lots of foundational changes to how physics and updates from the server are handled, on top of implementing a much more difficult physics networking scheme. Even with a well-written, bug-free PR, there would probably be concerns about maintainability in case the project no longer has developers competent enough to maintain the code. Such a networking model would also increase the network and computational requirements on the server, so depending on how much overhead the change adds, a PR might be required to allow Minetest to be configured to use the old networking (I don't think this itself would be super difficult, but it would mean maintaining two ways of doing networked physics, and the dev might feel that it's a pain not to be able to just blow away the old code).

The bare minimum to improve these issues I think would be to have the following:
* Timestamps are sent with client->server or server->client movement updates
* The client extrapolates entity movement prediction from when the update happened (currently the client just sets the CAO physics state to what it was when the update happened, which means the physics state may lag behind what the server has)
* The client sends enough information for the server to fully reconstruct player movement, in the time range that it has received information for. Probably the simplest and least bandwidth-using information would just be a log of player input states since the last update.
* The server predicts player movement and uses the timestamp to synchronize it properly like above

This would already be halfway to fully server-authoritative player movement, and would be enough to resolve the case where the player controls are not changing and no unpredictable physics state is being set on the entity (lua API calls or collisions that happen with parts of the map that change). The missing part for full server-side player movement is having the server correct the player's physics state. This half-implementation is already enough to have the downsides I mentioned earlier, so there wouldn't be any marginal benefit (other than easier development) over a full implementation.

**Comment by TheTermos** [comment]
> > upon receiving an update, the client attempts a move instead of setting a position.
> 
> The client already "attempts moves" (assuming you mean checks for collision when setting position, please clarify if you mean something else) when processing player movement.

The idea was about entities, obviously, players are already client controlled.

It would prevent an entity's box from clipping with player's client side, even if the server tells them to, that way player wouldn't be allowed to move through them because collision detection.
Client side collisions would take place in a slightly different spot than server side, but that's already the case.

> * Timestamps are sent with client->server or server->client movement updates
> * The client extrapolates entity movement prediction from when the update happened (currently the client just sets the CAO physics state to what it was when the update happened, which means the physics state may lag behind what the server has)

This should be done either way, if it isn't that would explain all the lag-like effects with moving entities happening even in local play.

**Comment by raymoo** [comment]
> The idea was about entities, obviously, players are already client controlled.

I see what you mean now, thanks for clarifying. This would probably require some way to differentiate between the entity position changing due to the lua API and just changing due to velocity changes, though (to allow entities to teleport even if there is something in the way). Also it could be more than "slightly out of sync", for example in the following situation after an entity has clipped through a player and hasn't stopped moving on the server side:
```
cp                    s

c - entity cao position
p - player position (on client)
s - entity sao position
```

I don't know if this is a large problem, my guess is that in most cases it will be a graphical annoyance + the player having to walk around. If it happens in a hallway it might block a player that could otherwise have progressed, but I'm not sure of a case where the player would be blocked with this change and not blocked if entity positions were perfectly synced.

**Comment by TheTermos** [comment]
> This would probably require some way to differentiate between the entity position changing due to the lua API and just changing due to velocity changes

Good point, there goes simplicity.

> Also it could be more than "slightly out of sync", for example in the following situation after an entity has clipped through a player and hasn't stopped moving on the server side:

Well, with default update interval of 0.09, player max velocity of 4n/s and not taking into account any network delay, theoretically the 'slightly' can get as much as 0.36n. That's already quite macroscopic, but it only gets worse with increasing velocity and lag and can reach multiple nodes.

But I must stress this level of desync is what has existed since the beginning,  the suggestion doesn't introduce any new inaccuracy itself, it just trades player position uncertainty for client side only entity position uncertainty with a bonus of preserving collision.

With entities controlled server side and players client side, players and entities are impossible to sync.

**Comment by raymoo** [comment]
>it just trades player position uncertainty for client side only entity position uncertainty with a bonus of preserving collision.

I don't think this reduces player position uncertainty. The client always knows its position (since it's authoritative), and the server only updates the player position when receiving an update from the client, which this suggestion doesn't change. As far as I can tell, it just preserves collisions on the client at the cost of making entity positions less synchronized. I don't know whether this is good or bad, it would depend on concrete instances of it affecting gameplay.

**Comment by TheTermos** [comment]
> I don't think this reduces player position uncertainty. The client always knows its position (since it's authoritative), and the server only updates the player position when receiving an update from the client, which this suggestion doesn't change

It does from gameplay pov. Take any head on player-entity collision, it may either fail or succeed based only on the timing relative to server updates, all else factors being equal. The player may end up in either of two different spots up to nodes away, that seems quite some uncertainty to me.

Anyway, that's mostly brainstorming, really it just should be done right, too bad it probably ain't gonna.

I understand the appeal of controlling player client side at early project stages when lone player walks over static terrain - it makes movement seem nice and smooth, but it's wrong, it just makes more advanced interactions with moving server controlled objects impossible.

**Comment by kno10** [comment]
While collision.cpp can compute collisions with object boxes, AFAICT they are assumed to be static.
Can you verify that this problem does not occur if one entity is static?
